File: pwa/src/cache/map-cache-manager.ts
/**
 * Map tile cache manager.
 * Seeds tiles by programmatically panning a Leaflet map across a grid.
 * Tiles are cached by the service worker (CacheFirst strategy on OSM URLs).
 *
 * Ported from MapCacheManager.kt — same 3×3 grid × 4 zoom levels approach.
 */

import type L from 'leaflet';

const CACHE_RADIUS_DEGREES = 0.15; // ~15km
const CACHE_ZOOM_LEVELS = [10, 12, 14, 16];
const GRID_SIZE = 3;
const PAN_DELAY_MS = 400; // Slightly longer than Android (300ms) for web

const STORAGE_KEY = 'mapCache';

interface CacheMeta {
  lat: number;
  lon: number;
  radius: number;
  complete: boolean;
}

function getCacheMeta(): CacheMeta | null {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    return raw ? JSON.parse(raw) : null;
  } catch {
    return null;
  }
}

function saveCacheMeta(meta: CacheMeta): void {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(meta));
}

/** Check if we have a tile cache covering the given location. */
export function hasCacheForLocation(lat: number, lon: number): boolean {
  const meta = getCacheMeta();
  if (!meta?.complete) return false;

  const margin = meta.radius * 0.3;
  return (
    Math.abs(lat - meta.lat) < meta.radius - margin &&
    Math.abs(lon - meta.lon) < meta.radius - margin
  );
}

/**
 * Seed the tile cache by panning the map across the surrounding area.
 * The service worker's CacheFirst strategy on OSM tile URLs handles storage.
 *
 * Reports progress via callback (0 to 1).
 */
export async function cacheMapArea(
  map: L.Map,
  lat: number,
  lon: number,
  onProgress?: (progress: number) => void,
): Promise<boolean> {
  try {
    const totalSteps = CACHE_ZOOM_LEVELS.length * GRID_SIZE * GRID_SIZE;
    let step = 0;

    const originalZoom = map.getZoom();
    const originalCenter = map.getCenter();

    for (const zoom of CACHE_ZOOM_LEVELS) {
      map.setZoom(zoom);

      for (let row = 0; row < GRID_SIZE; row++) {
        for (let col = 0; col < GRID_SIZE; col++) {
          const panLat =
            lat -
            CACHE_RADIUS_DEGREES +
            (2 * CACHE_RADIUS_DEGREES * row) / (GRID_SIZE - 1);
          const panLon =
            lon -
            CACHE_RADIUS_DEGREES +
            (2 * CACHE_RADIUS_DEGREES * col) / (GRID_SIZE - 1);

          map.setView([panLat, panLon], zoom, { animate: false });
          map.invalidateSize();

          step++;
          onProgress?.(step / totalSteps);

          await sleep(PAN_DELAY_MS);
        }
      }
    }

    // Restore original view
    map.setView(originalCenter, originalZoom, { animate: false });

    saveCacheMeta({
      lat,
      lon,
      radius: CACHE_RADIUS_DEGREES,
      complete: true,
    });

    return true;
  } catch {
    return false;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: app/src/main/java/no/naiv/tilfluktsrom/data/MapCacheManager.kt
package no.naiv.tilfluktsrom.data

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import org.osmdroid.util.GeoPoint
import org.osmdroid.views.MapView

/**
 * Manages offline map tile caching for the surrounding area.
 *
 * OSMDroid's SqlTileWriter automatically caches every tile the MapView loads.
 * We exploit this by programmatically panning the map across the surrounding area
 * at multiple zoom levels, which causes tiles to be fetched and cached passively.
 *
 * This approach respects OSM's tile usage policy (no bulk download) while still
 * building up an offline cache for the user's area.
 */
class MapCacheManager(private val context: Context) {

    companion object {
        private const val TAG = "MapCacheManager"
        private const val PREFS_NAME = "map_cache_prefs"
        private const val KEY_CACHED_LAT = "cached_center_lat"
        private const val KEY_CACHED_LON = "cached_center_lon"
        private const val KEY_CACHE_RADIUS = "cache_radius_km"
        private const val KEY_CACHE_COMPLETE = "cache_complete"

        private const val CACHE_RADIUS_DEGREES = 0.15  // ~15km

        // Zoom levels to cache: overview down to street level
        private val CACHE_ZOOM_LEVELS = intArrayOf(10, 12, 14, 16)

        // Grid points per axis per zoom level for panning
        private const val GRID_SIZE = 3
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /**
     * Check if we have a map cache that covers the given location.
     */
    fun hasCacheForLocation(latitude: Double, longitude: Double): Boolean {
        if (!prefs.getBoolean(KEY_CACHE_COMPLETE, false)) return false

        val cachedLat = Double.fromBits(prefs.getLong(KEY_CACHED_LAT, 0))
        val cachedLon = Double.fromBits(prefs.getLong(KEY_CACHED_LON, 0))
        val radius = prefs.getFloat(KEY_CACHE_RADIUS, 0f).toDouble()

        if (radius == 0.0) return false

        val margin = radius * 0.3
        return Math.abs(latitude - cachedLat) < (radius - margin) &&
               Math.abs(longitude - cachedLon) < (radius - margin)
    }

    /**
     * Seed the tile cache by panning the map across the area around the location.
     * OSMDroid's built-in SqlTileWriter caches each tile as it's loaded.
     *
     * Reports progress via callback (0.0 to 1.0).
     */
    suspend fun cacheMapArea(
        mapView: MapView,
        latitude: Double,
        longitude: Double,
        onProgress: (Float) -> Unit = {}
    ): Boolean = withContext(Dispatchers.Main) {
        try {
            Log.i(TAG, "Seeding tile cache for area around $latitude, $longitude")

            val totalSteps = CACHE_ZOOM_LEVELS.size * GRID_SIZE * GRID_SIZE
            var step = 0

            for (zoom in CACHE_ZOOM_LEVELS) {
                mapView.controller.setZoom(zoom.toDouble())

                // Pan across a grid of points covering the area
                for (row in 0 until GRID_SIZE) {
                    for (col in 0 until GRID_SIZE) {
                        val lat = latitude - CACHE_RADIUS_DEGREES +
                            (2 * CACHE_RADIUS_DEGREES * row) / (GRID_SIZE - 1)
                        val lon = longitude - CACHE_RADIUS_DEGREES +
                            (2 * CACHE_RADIUS_DEGREES * col) / (GRID_SIZE - 1)

                        mapView.controller.setCenter(GeoPoint(lat, lon))
                        // Force a layout pass so tiles are requested
                        mapView.invalidate()

                        step++
                        onProgress(step.toFloat() / totalSteps)

                        // Brief delay to allow tile loading to start
                        delay(300)
                    }
                }
            }

            // Restore to user's location
            mapView.controller.setZoom(14.0)
            mapView.controller.setCenter(GeoPoint(latitude, longitude))

            prefs.edit()
                .putLong(KEY_CACHED_LAT, latitude.toBits())
                .putLong(KEY_CACHED_LON, longitude.toBits())
                .putFloat(KEY_CACHE_RADIUS, CACHE_RADIUS_DEGREES.toFloat())
                .putBoolean(KEY_CACHE_COMPLETE, true)
                .apply()

            Log.i(TAG, "Tile cache seeding complete")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to seed tile cache", e)
            false
        }
    }
}
